### PR TITLE
Concurrency and Parallelism for Flix

### PR DESCRIPTION
**Pull request by jacobtonder**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 1]
```diff
@@ -707,6 +734,78 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
...
+          val loc = mkSL(sp1, sp2)
+          @@(visit(exp1, unsafe), visit(exp2, unsafe)) map {
+            case (e1, e2) => WeededAst.Expression.PutChannel(e1, e2, loc)
+          }
+
+        case ParsedAst.Expression.Spawn(sp1, fn, params, sp2) =>
+          val loc = mkSL(sp1, sp2)
+          val func = WeededAst.Expression.VarOrDef(Name.QName(sp1, Name.RootNS, fn, sp2), loc)
+
+          // TODO simplify this
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala]: I am okay if you cannot simplify it further, but could you add some more comments describing each step?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala]: Especially add comment that describe the overall goal and shows the transformation at a high level, i.e.
spawn f(e1, e2, e3) -> ...

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 2]
```diff
@@ -0,0 +1,705 @@
...
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.ExecutableAst.Root
+import org.objectweb.asm.Label
+import org.objectweb.asm.ClassWriter
+import org.objectweb.asm.Opcodes._
+
+/**
+  * Generates bytecode for the channel classes.
+  */
+object GenChannelClasses {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: Ensure that the code is formatted correctly. Use idea, Code -> Reformat Code.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+
+    // Generating each channel class
+    types.map{ tpe =>
+      val classType = JvmName.getChannelClassType(tpe)
+      classType.name -> JvmClass(classType.name, genChannelClass(classType, tpe))
+    }.toMap
+  }
+
+  /**
+    * Generating class `classType` with value of type `channelType`
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: Please add some Java pseudo code that shows the overall structure of the generated class or shows an example for a specific type.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+    *             this.signalPutters();
+    *         }
+    *     } finally {
+    *         this.unlock();
+    *     }
+    *
+    *     return value;
+    * }
+   */
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: Remove linebreak disconnecting the comment from the method.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 5]
```diff
@@ -0,0 +1,705 @@
...
+    mv.visitInsn(iRet)
+    mv.visitMaxs(4, 4)
+    mv.visitEnd()
+  }
+
+  /**
+    * Generates the `putValue()` method of the `classType` with value of type `channelType`
+    *
+    * public `classType` putValue(tpe value) throws InterruptedException {
+    *     this.lock.lock();
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: just this.lock() or?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 6]
```diff
@@ -0,0 +1,705 @@
...
+    *
+    *     try {
+    *         while(this.isFull()) {
+    *             this.awaitGetters();
+    *         }
+    *
+    *         this.offer(value);
+    *         this.signalGetters();
+    *     } finally {
+    *         this.lock.unlock();
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: this.unlock

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 7]
```diff
@@ -0,0 +1,705 @@
...
+    mv.visitJumpInsn(GOTO, loopStart)
+    mv.visitLabel(loopEnd)
+
+    // Offer
+    mv.visitVarInsn(ALOAD, 0)
+    mv.visitVarInsn(iLoad, 1)
+    mv.visitMethodInsn(INVOKEVIRTUAL, classType.name.toInternalName, "offer", AsmOps.getMethodDescriptor(List(channelType), JvmType.PrimBool), false)
+
+    // TODO: Use the variable instead of just popping it
+    mv.visitInsn(POP)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: Is something incomplete here?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 8]
```diff
@@ -0,0 +1,705 @@
...
+
+    // Signal All
+    mv.visitVarInsn(ALOAD, 0)
+    mv.visitMethodInsn(INVOKEVIRTUAL, classType.name.toInternalName, "signalGetters", AsmOps.getMethodDescriptor(Nil, JvmType.Void), false)
+
+    // TODO: Clear Selects
+    //mv.visitVarInsn(ALOAD, 0)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: What is happening here?

If selects are not cleared that could cause problems (and definitely memory leaks).

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 9]
```diff
@@ -0,0 +1,705 @@
...
+    val mv = visitor.visitMethod(ACC_PUBLIC, "putValue", AsmOps.getMethodDescriptor(List(channelType), classType), null, Array(JvmName.InterruptedException.toInternalName))
+    val labelStart = new Label()
+    val labelEnd = new Label()
+    val labelHandler = new Label()
+    val loopStart = new Label()
+    val loopEnd = new Label()
+    val labelReturn = new Label()
+    mv.visitCode()
+
+    // Lock
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: invoke your own lock method?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+    mv.visitFieldInsn(GETFIELD, classType.name.toInternalName, "lock", JvmType.Lock.toDescriptor)
+    mv.visitMethodInsn(INVOKEINTERFACE, JvmType.Lock.name.toInternalName, "unlock", AsmOps.getMethodDescriptor(Nil, JvmType.Void), true)
+
+    mv.visitJumpInsn(GOTO, labelReturn)
+
+    // Catch
+    mv.visitLabel(labelHandler)
+    mv.visitInsn(POP)
+
+    mv.visitVarInsn(ALOAD, 0)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala]: invoke your own unlock method?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenClosureClasses.scala [hunk 11]
```diff
@@ -110,12 +110,73 @@ object GenClosureClasses {
...
+    */
+  private def compileCopyMethod(visitor: ClassWriter, classType: JvmType.Reference, functionInterface: JvmType.Reference, defn: Def,
+                                 freeVars: List[FreeVar], resultType: JvmType)(implicit root: Root, flix: Flix): Unit = {
+    // Variable types
+    val varTypes = freeVars.map(_.tpe).map(JvmOps.getErasedJvmType)
+    // Function parameters
+    val params = defn.formals.takeRight(defn.formals.length - freeVars.length)
+
+    // Method header
+    val copyMethod = visitor.visitMethod(ACC_PUBLIC + ACC_FINAL, "copy",
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenClosureClasses.scala]: copyMethod -> mv

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 12]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
 
+    case Expression.NewChannel(exp, tpe, loc) =>
+      val classType = JvmOps.getChannelClassType(tpe)
+      // Adding source line number for debugging
+      addSourceLine(visitor, loc)
+      // Instantiating a new object of channel
+      visitor.visitTypeInsn(NEW, classType.name.toInternalName)
+      // Duplicating the object reference
+      visitor.visitInsn(DUP)
+      // Evaluate the underlying expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Evaluate the channel size expression.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 13]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+      // Constructor descriptor
+      val constructorDescriptor = AsmOps.getMethodDescriptor(List(JvmType.PrimInt), JvmType.Void)
+      // Call the constructor
+      visitor.visitMethodInsn(INVOKESPECIAL, classType.name.toInternalName, "<init>", constructorDescriptor, false)
+
+    case Expression.GetChannel(exp, tpe, loc) =>
+      val classType = JvmOps.getChannelClassType(exp.tpe)
+      // Adding source line number for debugging
+      addSourceLine(visitor, loc)
+      // Evaluate the underlying expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Evaluate the channel expression.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 14]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+      compileExpression(exp, visitor, currentClass, lenv0, entryPoint)
+      // Constructor descriptor
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: getValue descriptor

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 15]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+      val methodDescriptor = AsmOps.getMethodDescriptor(Nil, JvmOps.getErasedJvmType(tpe))
+      // Call the constructor
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Call getValue.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 16]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+      visitor.visitMethodInsn(INVOKEVIRTUAL, classType.name.toInternalName, "getValue", methodDescriptor, false)
+
+      if (tpe.typeConstructor.isChannel)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: This looks very weird. Is this not always the case? What happens if it is not?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 17]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+      {
+        val channelType = "Channel" + "$" + JvmOps.stringify(JvmOps.getErasedJvmType(tpe.typeArguments.head))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Add a helper method for this.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 18]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+        val jvmChannelType = JvmType.Reference(JvmName(List("ca", "uwaterloo", "flix"), channelType))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Add a helper method for this.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 19]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+
+        visitor.visitTypeInsn(CHECKCAST, jvmChannelType.name.toInternalName)
+      }
+
+    case Expression.PutChannel(exp1, exp2, tpe, loc) =>
+      val classType = JvmOps.getChannelClassType(exp1.tpe)
+
+      // Adding source line number for debugging
+      addSourceLine(visitor, loc)
+      // Evaluate the underlying expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Fix comments :)

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 20]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+      compileExpression(exp1, visitor, currentClass, lenv0, entryPoint)
+      // Evaluate the underlying expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Fix comments :)

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 21]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+      compileExpression(exp2, visitor, currentClass, lenv0, entryPoint)
+      // Constructor descriptor
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Fix comments :)

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 22]
```diff
@@ -816,6 +817,84 @@ object GenExpression {
...
+      val methodDescriptor = AsmOps.getMethodDescriptor(List(JvmOps.getErasedJvmType(exp2.tpe)), classType)
+      // Call the constructor
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Fix comments :)

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunctionInterfaces.scala [hunk 23]
```diff
@@ -89,6 +89,10 @@ object GenFunctionInterfaces {
       setArgMethod.visitEnd()
     }
 
+    val copyMethod = visitor.visitMethod(ACC_PUBLIC + ACC_ABSTRACT, "copy",
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunctionInterfaces.scala]: You dont have to store the result in the local variable copyMethod.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/JvmType.scala [hunk 24]
```diff
@@ -37,6 +37,57 @@ sealed trait JvmType {
...
+    case JvmType.Void => JvmType.Void
+    case JvmType.PrimBool => JvmType.Boolean
+    case JvmType.PrimChar => JvmType.Character
+    case JvmType.PrimByte => JvmType.Byte
+    case JvmType.PrimShort => JvmType.Short
+    case JvmType.PrimInt => JvmType.Integer
+    case JvmType.PrimLong => JvmType.Long
+    case JvmType.PrimFloat => JvmType.Float
+    case JvmType.PrimDouble => JvmType.Double
+    //case JvmType.Object => JvmType.Object
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/JvmType.scala]: Remove.

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala [hunk 25]
```diff
@@ -337,6 +392,180 @@ object Interpreter {
...
+      }
+
+      // Get value of channel.
+      value = chan.queue.poll()
+      if (value != null) {
+        // Signal all putters on the channel.
+        chan.channelGetters.signalAll()
+      }
+    }
+    finally {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala]: move up one line, } finally {

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala [hunk 26]
```diff
@@ -337,6 +392,180 @@ object Interpreter {
...
+
+      // Put value to channel.
+      chan.queue.add(value)
+
+      // Signal all getters & select conditions on the channel.
+      chan.channelPutters.signalAll()
+      signalAllConditions(chan.conditions)
+      // Clear select conditions
+      chan.conditions.clear()
+    }
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala]: move up one line, } finally {

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala [hunk 27]
```diff
@@ -337,6 +392,180 @@ object Interpreter {
...
+      // Unlock the channel.
+      chan.lock.unlock()
+    }
+
+    // Return the channel.
+    chan
+  }
+
+  /**
+   * Select channel
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala]: Document what it returns.

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala [hunk 28]
```diff
@@ -337,6 +392,180 @@ object Interpreter {
...
+   */
+  private def selectChannel(rules: List[(Symbol.VarSym, Value.Channel, Expression)]): (Symbol.VarSym, AnyRef, Expression) = {
+    val sLock = new ReentrantLock
+    val sCondition = sLock.newCondition
+    val channels = rules.map(_._2).sorted
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala]: Add a comment to emphasize that the sorted order is important.

File: main/test/ca/uwaterloo/flix/language/feature/FeatureSuite.scala [hunk 29]
```diff
@@ -32,6 +32,12 @@ class FeatureSuite extends Suites(
   new FlixTest("Test.Expression.LetMatch", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.LetMatch.flix"),
   new FlixTest("Test.Pattern.List", "main/test/ca/uwaterloo/flix/language/feature/Test.Pattern.List.flix"),
   new FlixTest("Test.Expression.LambdaMatch", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.LambdaMatch.flix"),
+  new FlixTest("TestStatement", "main/test/ca/uwaterloo/flix/language/feature/TestStatement.flix"),
+  new FlixTest("Test.Expression.NewChannel", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.NewChannel.flix"),
+  new FlixTest("Test.Expression.GetChannel", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.GetChannel.flix"),
+  new FlixTest("Test.Expression.PutChannel", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.PutChannel.flix"),
+  new FlixTest("Test.Expression.Spawn", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Spawn.flix"),
+  //new FlixTest("Test.Expression.SelectChannel", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.SelectChannel.flix"), // TODO
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/FeatureSuite.scala]: Tests not working?

**Comment by magnus-madsen** [comment]
I spent a lot of time reviewing the code. Overall its look really good. I pointed out a few important issues and lots of minor formatting and naming improvements.

I do have few other concerns:

1. What is the status of channels with buffer size of zero? Do they work? 
2. Is select fully implemented for the JVM backend? What is missing?
3. Does the JVM backend for channels generate unique ids as for the interpreter?

If something is still missing what are the prospects for (eventually) getting these things fixed?
I know you are busy with other stuff, but is there some interest in completely finishing the work?

**Comment by jacobtonder** [comment]
We are currently busy for the next 2 weeks, so expect little to no work being done.

To answer your questions

1. Currently channels with a buffer size of zero results in a deadlock, as it has not been implemented.
2. Selected is not implemented in the JVM backend.
3. Considering select has not been impleneted, no unique ids are being generated for channels in the JVM backend.

We plan on working on this before the project exam.

**Pull request closed without merging**